middleware/decrypt.py

```diff
@@ -0,0 +1,127 @@
+from crypt4gh.keys import get_private_key  # type: ignore
+
+
+def get_private_keys(file_paths: list[Path]) -> list[bytes]:
+    """Retrieve private keys from a list of files.
+
+    Args:
+        file_paths: A list of file paths.
+
+    Returns:
+        A list of retrieved private keys as byte objects.
+    """
+    private_keys = []
+    for file_path in file_paths:
+        try:
+            # Callback returns password of sk
+            key = get_private_key(filepath=file_path, callback=lambda x: '')
+            private_keys.append(key)
+        except ValueError:
+            continue
+    return private_keys
+
+
+def decrypt_files(file_paths: list[Path], private_keys: list[bytes]):
+    """Decrypt files in place.
+
+    Args:
+        file_paths: A list of file paths.
+        private_keys: A list of private keys as byte objects.
+
+    Raises:
+        ValueError: If no private key for a Crypt4GH file is provided.
+    """
+    encryption_method_codes = {
+        'ChaCha20': 0,
+        'AES-GCM': 1  # Not currently supported by Crypt4GH standard
+    }
+    # Third element of tuple is the recipient pk, which isn't used in decryption
+    key_tuples = [(encryption_method_codes['ChaCha20'], sk, None) for sk in private_keys]
+    for file_path in file_paths:
+        with open(file_path, "rb") as f_in, NamedTemporaryFile() as f_out:
+            try:
+                decrypt(keys=key_tuples, infile=f_in, outfile=f_out)  # Checks for magic
+                shutil.move(f_out.name, file_path)
+            except ValueError as e:
+                if str(e) != "Not a CRYPT4GH formatted file":
+                    print(f"Private key for {file_path} not provided")
+                continue
+
+
+def move_files(file_paths: list[Path], output_dir: Path) -> list[Path]:
+    """Move files to a specified output directory.
+
+    Args:
+        file_paths: A list of file paths with unique file names.
+        output_dir: Directory to move files to.
+
+    Returns:
+        A list containing the new file paths.
+    """
+    output_paths = []
+    existing_names = set()
+    for file_path in file_paths:
+        if file_path.name in existing_names:
+            raise ValueError(f"Duplicate file name found: {file_path.name}")
+        output_paths = [output_dir/file_path.name for file_path in file_paths]
+        existing_names.add(file_path.name)
+    for src, dest in zip(file_paths, output_paths):
+        shutil.move(src, dest)
+    return output_paths
+
+
+def remove_files(directory: Path):
+    """Rewrites and removes all files in a directory using rm -P.
+
+    Args:
+        directory: Directory that holds the files to be deleted.
+    """
+    if not directory.is_dir():
+        raise ValueError(f"Could not remove files: {directory} is not a directory.")
+    for file in directory.iterdir():
+        subprocess.run(["rm", "-P", str(file)], check=True)
+
+
+def main():
+    """Parse command-line arguments"""
```

Comment:
I think it does more than that - it coordinates the execution of the entire script. In fact, I still think you could externalize the CLI args parsing into its own function for clarity.

Generally, I try to follow the "<=10 statements" and "single purpose" rules for functions/methods. Of course there are exceptions, but usually they can be avoided. Your main function barely makes the first rule, but not the second (illustrated by your docstring - it parses CLI args AND it coordinates the execution of the entire script).